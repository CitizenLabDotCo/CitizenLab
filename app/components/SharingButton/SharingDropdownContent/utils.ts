import tracks from '../tracks';
import { trackEventByName } from 'utils/analytics';

export type UtmParams = {
  source: string;
  campaign: string;
  content?: string;
};

export type Medium =
  | 'facebook'
  | 'twitter'
  | 'messenger'
  | 'whatsapp'
  | 'email';

export function addUtmToUrl(
  medium: Medium,
  url: string,
  utmParams?: UtmParams
) {
  let resUrl = url;

  if (utmParams) {
    resUrl += `?utm_source=${utmParams.source}&utm_campaign=${utmParams.campaign}&utm_medium=${medium}`;

    if (utmParams.content) {
      resUrl += `&utm_content=${utmParams.content}`;
    }
  }

  return resUrl;
}

export function handleClick(medium: Medium, href?: string) {
  if (href) {
    window.location.href = href;
  }
  trackEventByName(tracks.clickShare.name, { network: medium });
}

export function trackClick(medium: Medium) {
  trackEventByName(tracks.clickShare.name, { network: medium });
}
