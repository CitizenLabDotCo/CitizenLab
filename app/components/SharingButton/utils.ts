export type UtmParams = {
  source: string;
  campaign: string;
  content?: string;
};

export type Medium =
  | 'facebook'
  | 'twitter'
  | 'messenger'
  | 'whatsapp'
  | 'email';

export function addUtmToUrl(medium: Medium, url: string, utmParams: UtmParams) {
  let resUrl = url;

  resUrl += `?utm_source=${utmParams.source}&utm_campaign=${utmParams.campaign}&utm_medium=${medium}`;

  if (utmParams.content) {
    resUrl += `&utm_content=${utmParams.content}`;
  }

  return resUrl;
}
