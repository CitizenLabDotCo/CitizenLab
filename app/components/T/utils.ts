import { makeSelectSetting } from 'utils/tenant/selectors';
import { makeSelectLocale } from 'containers/LanguageProvider/selectors';
import { Iterable, List } from 'immutable';
import { connect } from 'react-redux';
import { Multiloc, Locale } from 'typings';

const isImmutable = Iterable.isIterable;

const findTranslatedTextMutable = (value: Multiloc | undefined, userLocale: Locale, tenantLocales: List<Locale>): string => {
  let text = '';

  if (!value) {
    return text;
  }

  if (value[userLocale]) {
    text = value[userLocale] || '';
  } else {
    tenantLocales.some((tenantLocale) => {
      if (tenantLocale && value[tenantLocale]) {
        text = value[tenantLocale] || '';
        return true;
      }
      return false;
    });
  }
  return text;
};


function findTranslatedTextImmutable(value: Map<Locale, string>, userLocale: Locale, tenantLocales: List<Locale>): string {
  let text = '';

  if (!value) {
    return text;
  }

  if (value.get(userLocale)) {
    text = value.get(userLocale) || '';
  } else {
    tenantLocales.some((tenantLocale) => {
      if (tenantLocale && value.get(tenantLocale)) {
        text = value.get(tenantLocale) || '';
        return true;
      }
      return false;
    });
  }
  return text;
}

export const findTranslatedText = (value: Multiloc | Map<Locale, string> | undefined, userLocale: Locale, tenantLocales: List<Locale>) => {
  if (!isImmutable(value)) {
    return findTranslatedTextMutable(value as Multiloc, userLocale, tenantLocales);
  }
  return findTranslatedTextImmutable(value as Map<Locale, string>, userLocale, tenantLocales);
};


// Wrap a component with this function to make props.tFunc available. tFunc
// works as the T component, but as a function. This is useful for passing a
// translated value as a prop, e.g. as a placeholder
export function injectTFunc(component) {
  const mapStateToProps = (state) => {
    const selectLocale = makeSelectLocale();
    const selectTenantLocales = makeSelectSetting(['core', 'locales']);
    return {
      tFunc(value) {
        return findTranslatedText(value, selectLocale(state), selectTenantLocales(state));
      },
    };
  };

  return connect(mapStateToProps)(component);
}
