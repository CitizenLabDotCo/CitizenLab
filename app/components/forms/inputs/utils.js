
export const appenDableName = (name) => {
  const camelName = name.replace(/((_|^)\w)/g, (m) => m.slice(-1)[0].toUpperCase());
  return camelName.charAt(0).toUpperCase() + camelName.slice(1);
};

export const toCammelCase = (name) => name.replace(/(_\w)/g, (m) => m.slice(-1)[0].toUpperCase());
