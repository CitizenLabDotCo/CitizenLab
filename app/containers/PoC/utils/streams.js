import 'whatwg-fetch';
import Rx from 'rxjs/Rx';
import _ from 'lodash';
import request from 'utils/request';
import { v4 as uuid } from 'uuid';

class Streams {
  constructor() {
    this.listOfStreams = [];
  }

  create(apiEndpoint, queryParameters = null, localProperties = false) {
    const existingStream = this.listOfStreams.find((stream) => {
      return (
        _.isEqual(stream.apiEndpoint, apiEndpoint) &&
        _.isEqual(stream.queryParameters, queryParameters) &&
        _.isEqual(stream.localProperties, localProperties)
      );
    });

    if (!existingStream) {
      const newStream = {
        id: uuid(),
        apiEndpoint,
        queryParameters,
        localProperties,
        observer: null,
        observable: null,
        data: null,
      };

      newStream.observable = Rx.Observable.create((observer) => {
        newStream.observer = observer;

        request(apiEndpoint, null, null, queryParameters).then((response) => {
          observer.next(response.data);
        });

        return () => {
          this.listOfStreams = this.listOfStreams.filter((stream) => stream.id !== newStream.id);
        };
      })
      .startWith('initial')
      .scan((accumulated, current) => {
        let data = accumulated;

        if (!_.isFunction(current) && _.isObject(localProperties)) {
          if (_.isArray(current)) {
            data = current.map((child) => ({ ...child, ...localProperties }));
          } else if (_.isObject(current)) {
            data = { ...current, ...localProperties };
          } else {
            console.log('current is no Object or Array');
          }
        } else if (_.isFunction(current)) {
          data = current(data);
        } else {
          data = current;
        }

        newStream.data = data;

        return data;
      })
      .filter((data) => data !== 'initial')
      .distinctUntilChanged()
      .publishReplay(1)
      .refCount();

      this.listOfStreams = [...this.listOfStreams, newStream];

      return newStream;
    }

    return existingStream;
  }
}

export default new Streams();
