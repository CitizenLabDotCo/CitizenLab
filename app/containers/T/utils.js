import { makeSelectSetting } from 'utils/tenant/selectors';
import { makeSelectLocale } from 'containers/LanguageProvider/selectors';
import { Iterable } from 'immutable';
import { connect } from 'react-redux';

const isImmutable = Iterable.isIterable;

const findTranslatedTextMutable = (value, userLocale, tenantLocales) => {
  let text = '';

  if (value && value[userLocale]) {
    text = value[userLocale];
  } else {
    tenantLocales.some((tenantLocale) => {
      if (value && value[tenantLocale]) {
        text = value[tenantLocale];
        return true;
      }
      return false;
    });
  }
  return text;
};


function findTranslatedTextImmutable(value, userLocale, tenantLocales) {
  let text = '';
  if (value && value.get(userLocale)) {
    text = value.get(userLocale);
  } else {
    tenantLocales.some((tenantLocale) => {
      if (value && value.get(tenantLocale)) {
        text = value.get(tenantLocale);
        return true;
      }
      return false;
    });
  }
  return text;
}

export const findTranslatedText = (value, userLocale, tenantLocales) => {
  if (!isImmutable(value)) {
    // console.log(new Error('Attention, Data Provided to the T component should be immutable!'));
    return findTranslatedTextMutable(value, userLocale, tenantLocales);
  }
  return findTranslatedTextImmutable(value, userLocale, tenantLocales);
};


// Wrap a component with this function to make props.tFunc available. tFunc
// works as the T component, but as a function. This is useful for passing a
// translated value as a prop, e.g. as a placeholder
export function injectTFunc(component) {
  const mapStateToProps = (state) => {
    const selectLocale = makeSelectLocale();
    const selectTenantLocales = makeSelectSetting(['core', 'locales']);
    return {
      tFunc(value) {
        return findTranslatedText(value, selectLocale(state), selectTenantLocales(state));
      },
    };
  };

  return connect(mapStateToProps)(component);
}
