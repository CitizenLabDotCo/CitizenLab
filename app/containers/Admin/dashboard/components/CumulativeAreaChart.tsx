import React from 'react';
import { Subscription } from 'rxjs';
import { map, isNumber } from 'lodash-es';
import { FormattedMessage, injectIntl } from 'utils/cl-intl';
import { InjectedIntlProps } from 'react-intl';
import { withTheme } from 'styled-components';
import { AreaChart, Area, Tooltip, XAxis, YAxis, ResponsiveContainer } from 'recharts';
import messages from '../messages';

// components
import EmptyGraph from './EmptyGraph';
import { GraphCard, GraphCardInner, GraphCardTitle, GraphCardFigureContainer, GraphCardFigure, GraphCardFigureChange } from '../';

type State = {
  serie: {
    name: string | number,
    value: number,
    code: string
  }[] | null;
};

type Props = {
  className?: string;
  graphUnit: 'ActiveUsers' | 'Users' | 'Ideas' | 'Comments' | 'Votes';
  graphTitleMessageKey: string;
  startAt: string;
  endAt: string;
  resolution: 'month' | 'day';
  currentProjectFilter: string | null;
  currentGroupFilter: string | null;
  currentTopicFilter: string | null;
  stream: Function;
};

class CumulativeAreaChart extends React.PureComponent<Props & InjectedIntlProps, State> {
  subscription: Subscription;

  constructor(props: Props & InjectedIntlProps) {
    super(props as any);
    this.state = {
      serie: null,
    };
  }

  componentDidMount() {
    const {
      startAt,
      endAt,
      resolution,
      currentGroupFilter,
      currentTopicFilter,
      currentProjectFilter,
    } = this.props;

    this.resubscribe(
      startAt,
      endAt,
      resolution,
      currentGroupFilter,
      currentTopicFilter,
      currentProjectFilter,
    );
  }

  componentDidUpdate(prevProps: Props) {
    const {
      startAt,
      endAt,
      resolution,
      currentGroupFilter,
      currentTopicFilter,
      currentProjectFilter,
    } = this.props;

    if (startAt !== prevProps.startAt
      || endAt !== prevProps.endAt
      || resolution !== prevProps.resolution
      || currentGroupFilter !== prevProps.currentGroupFilter
      || currentTopicFilter !== prevProps.currentTopicFilter
      || currentProjectFilter !== prevProps.currentProjectFilter
    ) {
      this.resubscribe(
        startAt,
        endAt,
        resolution,
        currentGroupFilter,
        currentTopicFilter,
        currentProjectFilter,
      );
    }
  }

  componentWillUnmount() {
    this.subscription.unsubscribe();
  }

  convertToGraphFormat = (serie: { [key: string]: number }) => {
    return map(serie, (value, key) => ({
      value,
      name: key,
      code: key
    }));
  }

  resubscribe(
    startAt: string,
    endAt: string,
    resolution: 'month' | 'day',
    currentGroupFilter: string | null,
    currentTopicFilter: string | null,
    currentProjectFilter: string | null,
  ) {
    const { stream } = this.props;

    if (this.subscription) {
      this.subscription.unsubscribe();
    }

    this.subscription = stream({
      queryParameters: {
        start_at: startAt,
        end_at: endAt,
        interval: resolution,
        project: currentProjectFilter,
        group: currentGroupFilter,
        topic: currentTopicFilter,
      }
    }).observable.subscribe((serie) => {
      const convertedSerie = this.convertToGraphFormat(serie);
      this.setState({ serie: convertedSerie });
    });
  }

  formatTick = (date: string) => {
    const { resolution } = this.props;
    const { formatDate } = this.props.intl;

    return formatDate(date, {
      day: (resolution === 'month' ? undefined : '2-digit'),
      month: 'short',
    });
  }

  formatLabel = (date: string) => {
    const { resolution } = this.props;
    const { formatDate } = this.props.intl;

    return formatDate(date, {
      day: (resolution === 'month' ? undefined : '2-digit'),
      month: 'long',
      year: 'numeric'
    });
  }

  formatSerieChange = (serieChange: number) => {
    if (serieChange > 0) {
      return `(+${serieChange.toString()})`;
    } else if (serieChange < 0) {
      return `(${serieChange.toString()})`;
    }
    return null;
  }

  render() {
    const { formatMessage } = this.props.intl;
    const { className, graphTitleMessageKey, graphUnit } = this.props;
    const { serie } = this.state;
    const isEmpty = !serie || serie.every(item => item.value === 0);
    const { chartFill, chartLabelSize, chartLabelColor, chartStroke } = this.props['theme'];
    const firstSerieValue = serie && serie[0].value;
    const lastSerieValue = serie && serie[serie.length - 1].value;
    const serieChange = isNumber(firstSerieValue) && isNumber(lastSerieValue) && (lastSerieValue - firstSerieValue);
    const formattedSerieChange = isNumber(serieChange) ? this.formatSerieChange(serieChange) : null;

    return (
      <GraphCard className={className}>
        {!isEmpty ?
          <GraphCardInner>
            <GraphCardTitle>
              <FormattedMessage {...messages[graphTitleMessageKey]} />
              <GraphCardFigureContainer>
                <GraphCardFigure>
                  {lastSerieValue}
                </GraphCardFigure>
                <GraphCardFigureChange>
                  {formattedSerieChange}
                </GraphCardFigureChange>
              </GraphCardFigureContainer>
            </GraphCardTitle>
            <ResponsiveContainer>
              <AreaChart data={serie}>
                <Area
                  type="monotone"
                  dataKey="value"
                  name={formatMessage(messages[`numberOf${graphUnit}`])}
                  dot={false}
                  fill={chartFill}
                  fillOpacity={1}
                  stroke={chartStroke}
                />
                <XAxis
                  dataKey="name"
                  interval="preserveStartEnd"
                  stroke={chartLabelColor}
                  fontSize={chartLabelSize}
                  tick={{ transform: 'translate(0, 7)' }}
                  tickFormatter={this.formatTick}
                />
                <YAxis
                  stroke={chartLabelColor}
                  fontSize={chartLabelSize}
                />
                <Tooltip
                  isAnimationActive={false}
                  labelFormatter={this.formatLabel}
                />
              </AreaChart>
            </ResponsiveContainer>
          </GraphCardInner>
        :
          <EmptyGraph unit={graphUnit} />
        }
      </GraphCard>
    );
  }
}

export default injectIntl<Props>(withTheme(CumulativeAreaChart as any) as any);
