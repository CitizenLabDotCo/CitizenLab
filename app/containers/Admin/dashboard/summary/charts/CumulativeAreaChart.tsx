// libraries
import React, { PureComponent } from 'react';
import { Subscription } from 'rxjs';
import { map, isEmpty } from 'lodash-es';

// intl
import { FormattedMessage, injectIntl } from 'utils/cl-intl';
import { InjectedIntlProps } from 'react-intl';
import messages from '../../messages';
import { requestBlob } from 'utils/request';
import { API_PATH } from 'containers/App/constants';
import { reportError } from 'utils/loggingUtils';
import { saveAs } from 'file-saver';

// styling
import styled, { withTheme } from 'styled-components';
import { rgba } from 'polished';
import { fontSizes } from 'utils/styleUtils';

// components
import {
  AreaChart,
  CartesianGrid,
  Area,
  Tooltip,
  XAxis,
  YAxis,
  ResponsiveContainer,
} from 'recharts';
import {
  IGraphUnit,
  IResolution,
  GraphCard,
  NoDataContainer,
  GraphCardInner,
  GraphCardHeader,
  GraphCardTitle,
  GraphCardFigureContainer,
  GraphCardFigure,
  GraphCardFigureChange,
} from '../..';
import Button from 'components/UI/Button';
import { Dropdown } from 'cl2-component-library';

// typings
import { IStreamParams, IStream } from 'utils/streams';
import { IUsersByTime, IIdeasByTime, ICommentsByTime } from 'services/stats';
import { IGraphFormat } from 'typings';

const DropdownButton = styled(Button)``;

const Container = styled.div`
  display: flex;
  align-items: end;
  position: relative;
  cursor: pointer;
`;

type State = {
  serie: IGraphFormat | null;
  exporting: boolean;
  dropdownOpened: boolean;
};

type IResourceByTime = IUsersByTime | IIdeasByTime | ICommentsByTime;

type Props = {
  className?: string;
  graphUnit: IGraphUnit;
  graphTitleMessageKey: string;
  startAt: string | null | undefined;
  endAt: string | null;
  resolution: IResolution;
  currentProjectFilter: string | undefined;
  currentGroupFilter: string | undefined;
  currentTopicFilter: string | undefined;
  stream: (streamParams?: IStreamParams | null) => IStream<IResourceByTime>;
  onDownloadSvg: (name: string, ref: any) => void;
};

export class CumulativeAreaChart extends PureComponent<
  Props & InjectedIntlProps,
  State
> {
  subscription: Subscription;
  currentChart = React.createRef<HTMLDivElement>();

  constructor(props: Props & InjectedIntlProps) {
    super(props as any);
    this.state = {
      serie: null,
      exporting: false,
      dropdownOpened: false,
    };
  }

  componentDidMount() {
    const {
      startAt,
      endAt,
      resolution,
      currentGroupFilter,
      currentTopicFilter,
      currentProjectFilter,
    } = this.props;

    this.resubscribe(
      startAt,
      endAt,
      resolution,
      currentGroupFilter,
      currentTopicFilter,
      currentProjectFilter
    );
  }

  componentDidUpdate(prevProps: Props) {
    const {
      startAt,
      endAt,
      resolution,
      currentGroupFilter,
      currentTopicFilter,
      currentProjectFilter,
    } = this.props;

    if (
      startAt !== prevProps.startAt ||
      endAt !== prevProps.endAt ||
      resolution !== prevProps.resolution ||
      currentGroupFilter !== prevProps.currentGroupFilter ||
      currentTopicFilter !== prevProps.currentTopicFilter ||
      currentProjectFilter !== prevProps.currentProjectFilter
    ) {
      this.resubscribe(
        startAt,
        endAt,
        resolution,
        currentGroupFilter,
        currentTopicFilter,
        currentProjectFilter
      );
    }
  }

  componentWillUnmount() {
    this.subscription.unsubscribe();
  }

  convertToGraphFormat = (data: IResourceByTime) => {
    const { graphUnit } = this.props;

    if (!isEmpty(data.series[graphUnit])) {
      return map(data.series[graphUnit], (value, key) => ({
        value,
        name: key,
        code: key,
      }));
    }

    return null;
  };

  resubscribe(
    startAt: string | null | undefined,
    endAt: string | null,
    resolution: IResolution,
    currentGroupFilter: string | undefined,
    currentTopicFilter: string | undefined,
    currentProjectFilter: string | undefined
  ) {
    const { stream } = this.props;

    if (this.subscription) {
      this.subscription.unsubscribe();
    }

    this.subscription = stream({
      queryParameters: {
        start_at: startAt,
        end_at: endAt,
        interval: resolution,
        project: currentProjectFilter,
        group: currentGroupFilter,
        topic: currentTopicFilter,
      },
    }).observable.subscribe((serie) => {
      const convertedSerie = this.convertToGraphFormat(serie);
      this.setState({ serie: convertedSerie });
    });
  }

  formatTick = (date: string) => {
    const { resolution } = this.props;
    const { formatDate } = this.props.intl;

    return formatDate(date, {
      day: resolution === 'month' ? undefined : '2-digit',
      month: 'short',
    });
  };

  formatLabel = (date: string) => {
    const { resolution } = this.props;
    const { formatDate } = this.props.intl;

    return formatDate(date, {
      day: resolution === 'month' ? undefined : '2-digit',
      month: 'long',
      year: 'numeric',
    });
  };

  formatSerieChange = (serieChange: number) => {
    if (serieChange > 0) {
      return `(+${serieChange.toString()})`;
    } else if (serieChange < 0) {
      return `(${serieChange.toString()})`;
    }
    return null;
  };

  getFormattedNumbers(serie: IGraphFormat | null) {
    if (serie) {
      const firstSerieValue = serie[0].value;
      const lastSerieValue = serie[serie.length - 1].value;
      const serieChange = lastSerieValue - firstSerieValue;
      let typeOfChange: 'increase' | 'decrease' | '' = '';

      if (serieChange > 0) {
        typeOfChange = 'increase';
      } else if (serieChange < 0) {
        typeOfChange = 'decrease';
      }

      return {
        typeOfChange,
        totalNumber: lastSerieValue,
        formattedSerieChange: this.formatSerieChange(serieChange),
      };
    }

    return {
      totalNumber: null,
      formattedSerieChange: null,
      typeOfChange: '',
    };
  }

  downloadXlsx = async () => {
    // const { exportQueryParameter } = this.props;
    const {
      startAt,
      endAt,
      resolution,
      currentGroupFilter,
      currentTopicFilter,
      currentProjectFilter,
    } = this.props;

    const queryParametersObject = {
      start_at: startAt,
      end_at: endAt,
      interval: resolution,
      project: currentProjectFilter,
      group: currentGroupFilter,
      topic: currentTopicFilter,
    };
    // if (isString(exportQueryParameter) && exportQueryParameter !== 'all') {
    //   queryParametersObject['project'] = exportQueryParameter;
    // } else if (!isString(exportQueryParameter)) {
    //   queryParametersObject['ideas'] = exportQueryParameter;
    // }

    try {
      this.setState({ exporting: true });
      const blob = await requestBlob(
        `${API_PATH}/stats/users_by_time_cumulative_as_xlsx`,
        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        queryParametersObject
      );
      saveAs(blob, 'ideas-export.xlsx');
      this.setState({ exporting: false });
    } catch (error) {
      reportError(error);
      this.setState({ exporting: false });
    }

    // track this click for user analytics
    // trackEventByName(tracks.clickExportIdeas.name);
  };

  removeFocus = (event: React.MouseEvent) => {
    event.preventDefault();
  };

  toggleDropdown = (event: React.FormEvent<any>) => {
    event.preventDefault();
    this.setState(({ dropdownOpened }) => ({
      dropdownOpened: !dropdownOpened,
    }));
  };

  render() {
    const {
      graphTitleMessageKey,
      graphUnit,
      className,
      intl: { formatMessage },
    } = this.props;
    const { serie, exporting, dropdownOpened } = this.state;
    const {
      chartFill,
      chartLabelSize,
      chartLabelColor,
      chartStroke,
      animationBegin,
      animationDuration,
    } = this.props['theme'];
    const formattedNumbers = this.getFormattedNumbers(serie);
    const {
      totalNumber,
      formattedSerieChange,
      typeOfChange,
    } = formattedNumbers;

    return (
      <GraphCard className={className}>
        <GraphCardInner>
          <GraphCardHeader>
            <GraphCardTitle>
              <FormattedMessage {...messages[graphTitleMessageKey]} />
            </GraphCardTitle>
            <GraphCardFigureContainer>
              <GraphCardFigure>{totalNumber}</GraphCardFigure>
              <GraphCardFigureChange className={typeOfChange}>
                {formattedSerieChange}
              </GraphCardFigureChange>
            </GraphCardFigureContainer>
            <Container className={className}>
              <DropdownButton
                buttonStyle="admin-dark-text"
                onClick={this.toggleDropdown}
                icon="download"
                iconPos="right"
                padding="0px"
              />
              <Dropdown
                width="100%"
                top="35px"
                right="-5px"
                mobileRight="-5px"
                opened={dropdownOpened}
                onClickOutside={this.toggleDropdown}
                content={
                  <>
                    <Button
                      onClick={() =>
                        this.props.onDownloadSvg(
                          formatMessage(messages[graphTitleMessageKey]),
                          this.currentChart
                        )
                      }
                      buttonStyle="text"
                      processing={exporting}
                      padding="0"
                      fontSize={`${fontSizes.small}px`}
                    >
                      Download svg
                    </Button>
                    <Button
                      onClick={this.downloadXlsx}
                      buttonStyle="text"
                      processing={exporting}
                      padding="0"
                      fontSize={`${fontSizes.small}px`}
                    >
                      Download xls
                    </Button>
                  </>
                }
              />
            </Container>
          </GraphCardHeader>
          {!serie ? (
            <NoDataContainer>
              <FormattedMessage {...messages.noData} />
            </NoDataContainer>
          ) : (
            <>
              <ResponsiveContainer>
                <AreaChart
                  data={serie}
                  margin={{ right: 40 }}
                  id="currentChart"
                  ref={(chart) =>
                    (this.currentChart = chart?.container?.children[0])
                  }
                >
                  <CartesianGrid strokeDasharray="5 5" />
                  <Area
                    type="monotone"
                    dataKey="value"
                    name={formatMessage(messages[graphUnit])}
                    dot={false}
                    fill={rgba(chartFill, 0.25)}
                    fillOpacity={1}
                    stroke={chartStroke}
                    animationDuration={animationDuration}
                    animationBegin={animationBegin}
                  />
                  <XAxis
                    dataKey="name"
                    interval="preserveStartEnd"
                    stroke={chartLabelColor}
                    fontSize={chartLabelSize}
                    tick={{ transform: 'translate(0, 7)' }}
                    tickFormatter={this.formatTick}
                  />
                  <YAxis stroke={chartLabelColor} fontSize={chartLabelSize} />
                  <Tooltip
                    isAnimationActive={false}
                    labelFormatter={this.formatLabel}
                  />
                </AreaChart>
              </ResponsiveContainer>
            </>
          )}
        </GraphCardInner>
      </GraphCard>
    );
  }
}

export default injectIntl<Props>(withTheme(CumulativeAreaChart as any) as any);
