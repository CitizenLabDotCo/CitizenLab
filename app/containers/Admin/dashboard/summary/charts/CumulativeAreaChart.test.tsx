import React from 'react';
import { shallow } from 'enzyme';
jest.mock('utils/cl-intl');
jest.mock('services/stats');
import { intl } from 'utils/cl-intl';
import { CumulativeAreaChart } from './CumulativeAreaChart';
import { chartTheme } from '../..';

const stream = jest.fn();
const serie = [
  {
    name: 'x',
    value: 1,
    code: 'x'
  },
  {
    name: 'y',
    value: 3,
    code: 'y'
  }
];

describe('<CumulativeAreaChart />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <CumulativeAreaChart
        intl={intl}
        theme={chartTheme({})}
        serie={serie}
        graphTitleMessageKey="commentsByTimeTitle"
        graphUnit="comments"
        startAt={'05-12-2018'}
        endAt={'06-12-2018'}
        resolution={'day'}
        stream={stream}
        {...this.state}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
