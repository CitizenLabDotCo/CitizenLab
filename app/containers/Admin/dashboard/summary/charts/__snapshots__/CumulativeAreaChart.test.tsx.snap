// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CumulativeAreaChart /> renders correctly with data 1`] = `
<dashboard__GraphCard>
  <dashboard__GraphCardInner>
    <dashboard__GraphCardHeader>
      <dashboard__GraphCardTitle>
        commentsByTimeTitle
      </dashboard__GraphCardTitle>
      <dashboard__GraphCardFigureContainer>
        <dashboard__GraphCardFigure />
        <dashboard__GraphCardFigureChange
          className=""
        />
      </dashboard__GraphCardFigureContainer>
    </dashboard__GraphCardHeader>
    <dashboard__NoDataContainer>
      <FormattedMessage
        defaultMessage="No data available with the current filters."
        id="app.containers.AdminPage.DashboardPage.noData"
      />
    </dashboard__NoDataContainer>
  </dashboard__GraphCardInner>
</dashboard__GraphCard>
`;

exports[`<CumulativeAreaChart /> renders correctly without data 1`] = `
<dashboard__GraphCard>
  <dashboard__GraphCardInner>
    <dashboard__GraphCardHeader>
      <dashboard__GraphCardTitle>
        commentsByTimeTitle
      </dashboard__GraphCardTitle>
      <dashboard__GraphCardFigureContainer>
        <dashboard__GraphCardFigure />
        <dashboard__GraphCardFigureChange
          className=""
        />
      </dashboard__GraphCardFigureContainer>
    </dashboard__GraphCardHeader>
    <dashboard__NoDataContainer>
      <FormattedMessage
        defaultMessage="No data available with the current filters."
        id="app.containers.AdminPage.DashboardPage.noData"
      />
    </dashboard__NoDataContainer>
  </dashboard__GraphCardInner>
</dashboard__GraphCard>
`;
